### ISSUE TITLE
Support / roadmap for typed Collections

### ISSUE THREAD

**Issue by holtkamp** [laravel/ideas#529]
Are there any plans to provide initial support for typed Collections? 

https://github.com/phpsafari/typed-collections seems like an effort. 

When the minimum requirement reaches PHP 7.0, an approach as described in https://github.com/phpsafari/typed-collections/issues/1 becomes available. A limitation of that approach would be that arrays/collections using a string as key are not supported by variadic parameters.

**Comment by matijakovacevic** [comment]
@holtkamp can you tell me what was the answer why this was closed?

I'm looking into
https://github.com/jeromegamez/typed-collection

I would still add to the above, the possibility to create typed collections dynamically
e.g. 

```php
$member = new Person('John', 'Doe');
$peopleCollection = new Collection([], Person::class);

$peopleCollection->push(new Person());
$peopleCollection->push($member);
```

**Comment by holtkamp** [comment]
>can you tell me what was the answer why this was closed?

Mmm, it seems it was never answered, not sure why it was closed.

Eventually I went with the approach as described in https://github.com/OveD-php/typed-collections/issues/1

In a domain model, typed collections like `Users`, `Invoices`, `InvoiceLines`, etc have a lot of benefits.

This approach *does* have the downside that each typed collection requires its own class / can not be created dynamically, but personally I never encountered this. As soon as you feel the need for a typed collection, it is usually worth making it explicit in a specific class. You will find out that it is nice to move functionality to that specific class.

A downside that I *did* encounter: when using `map()` on the typed collection, a collection of the same type is returned. So something like `$invoiceLines->getTotalValue()` can not be implemented using `map()`, but requires a `reduce()` operation feeding it a normal Collection. Of course this makes sense, can be annoying though ;)

Just my 2 cents :)
